### DISCUSSION TITLE
Persistent links for code line ranges?

### DISCUSSION THREAD

**Discussion by hellt** [squidfunk/mkdocs-material#5083] [Q&A] [answered]
It appears that code line ranges use the code block ID as they appear on the page when users select line ranges.

An issue with that approach is that when a new code block is added, the blocks under it get a new block ID, and hence the old links that were used to those code blocks are no longer valid.

I wonder, is it fair to suggest generating unique IDs for code blocks to alleviate this issue?

**Comment by squidfunk** [comment, marked as answer]
Perfectly reasonable request. For this to work, we need to tell the [Highlight](https://facelessuser.github.io/pymdown-extensions/extensions/highlight/) extension to allow for persistent code block ids. Thus, this would need to be supported upstream in [Python Markdown Extensions](https://facelessuser.github.io/pymdown-extensions/). Envisioned sample use:

```` markdown
``` .py id="foobar"
print("Hello world")
``` 
````

**Reply by facelessuser** [reply]
For others that are curious, we already support this with Pygments. You must have `attr_list` enabled.

````
```{.py #id}
import foo
```
````

**Reply by squidfunk** [reply]
Does this integrate with line anchors? The problem is, that we need the id to be part of the line anchor (and ideally line span and other elements), e.g., `__codelineno-foobar-10` instead of `__codelineno-0-10`. Otherwise, permalinks to line ranges won't work. If I'm not missing something, given that ids must be unique within a document, taking the id as a replacement for the sequence number should be reasonable.

**Reply by facelessuser** [reply]
Bah, I'll have to look. I think the question, when posed to me, wasn't very clear in the other issue. There may need to be some slight work there.

**Reply by facelessuser** [reply]
I've located what needs to change. It's a very, very small change. I'll get it out in the next release.

**Reply by squidfunk** [reply]
Awesome!

**Reply by squidfunk** [reply]
... please drop me a note when this is supported, we will update the docs then and recommend setting an identifier.

**Reply by facelessuser** [reply]
I'll try to remember 🙃. I imagine the next release you'll be aware of as formally we'll be adding the general purpose blocks publically. I imagine that is when I will formally poke about some of that support (collapsible code blocks, etc.). I just need to make sure the beta is relatively sound. With that said, I've migrated all Tabs, Admonitions, Definition, Details, and arbitrary HTML blocks in my documentation and it works quite well. So, this change will be in that release. I need to look over what else I want to pull into that incoming release.
